lib/cfg.js

```diff
@@ -0,0 +1,368 @@
+            rangeBb.actionPos = [];
+    };
+
+    while(first != last) {
+        var inst = asmArr[first].text;
+        if(isEnd(inst)) {
+            rangeBb.end = first;
+            result.push(JSON.parse(JSON.stringify(rangeBb)));
+            ++first;//risk of a bug.
+            //inst is expected to be .L*: where * in 1,2,...
+            resetRangeWith(rangeBb, inst, first);
+        }
+        else if(isJmp(inst)) {
+                rangeBb.actionPos.push(first);
+        }
+        ++first;
+    }
+
+    rangeBb.end = last;
+    result.push(JSON.parse(JSON.stringify(rangeBb)));
+    return result;
+};
+
+ControlFlowGraph.prototype.splitToCanonicalBasicBlock = function (basicBlock) {
+    var actionPos = basicBlock.actionPos;
+    var actPosSz = actionPos.length;
+    if (actionPos[actPosSz-1] + 1 == basicBlock.end){
+        --actPosSz;
+    }
+
+    if(actPosSz === 0)
+        return [{
+                    nameId: basicBlock.nameId,
+                    start: basicBlock.start,
+                    end: basicBlock.end
+               }];
+    else if(actPosSz == 1)
+            return [
+                {nameId: basicBlock.nameId,start: basicBlock.start,end:actionPos[0]+1},
+                {nameId: basicBlock.nameId+"@"+ (actionPos[0]+1),start: actionPos[0]+1,end: basicBlock.end}
+               ];
+    else {
+        var first = 0;
+        var last = actPosSz;
+        var block_name = basicBlock.nameId;
+        var tmp = {nameId:block_name, start:basicBlock.start, end:actionPos[first]+1 };
+        var result = [];
+        result.push(JSON.parse(JSON.stringify(tmp)));
+        while(first != last-1) {
+            tmp.nameId = block_name + "@" + (actionPos[first]+1);
+            tmp.start = actionPos[first]+1;
+            ++first;
+            tmp.end = actionPos[first]+1;
+            result.push(JSON.parse(JSON.stringify(tmp)));
+        }
+        
+        tmp = {nameId:block_name + "@" + (actionPos[first] + 1), start:actionPos[first] + 1, end:basicBlock.end };
+        result.push(JSON.parse(JSON.stringify(tmp)));
+        
+        return result;
+
+    }
+
+};
+
+ControlFlowGraph.prototype.concatInstructions = function (asmArr, first, last) {
+    
+    var makeShorter = function(instruction){
+        if(instruction.length > 50){
+            return instruction.substr(0, 50);
+        }
+        
+        return instruction;
+    };
+    
+    if(first == last) return "";//if last -1 is changed to last this line is no longuer needed
+    console.log("look");
+    var result = "";
+    //TODO: include \n for long templated instructions
+    while(first != last-1) {//added to delete last \n and handle the last concat outside loop
+        console.log(asmArr[first].text);
+        result += makeShorter(asmArr[first].text) + "\n";
+        ++first;
+    }
+    //last concat withou \n
+    result += makeShorter(asmArr[first].text);
+    console.log(asmArr[first].text);
+
+    return result;
+};
```

Comment:
I think the whole routine can be written as:

```
ControlFlowGraph.prototype.concatInstructions = function (asmArr, first, last) {
  return _.chain(asmArr.slice(first, last))
    .map(function (x) { return x.text.substr(0, 50); }
    .value()
    .join("\n");
};
```

Reply:
done. worked perfectly.